Cql/Cql.Runtime/ValueSets/HashValueSetDictionary.cs

```diff
@@ -3,101 +3,85 @@
-            if (!_codesInValueSet.TryGetValue(valueSetUri, out var codes))
-            {
-                codes = new HashSet<CqlCode>(_codeHasher) { code };
-                _codesInValueSet.Add(valueSetUri, codes);
-            }
-            else
-                codes.Add(code);
+        /// <summary>
+        /// Creates a new <see cref="HashValueSetDictionary"/> with the given <paramref name="comparer"/>.
+        /// </summary>
+        public HashValueSetDictionary(ICqlComparer<CqlCode> comparer)
+        {
+            _comparer = comparer;
+            _internHash = new ConcurrentDictionary<CqlCode, CqlCode>(_comparer.ToEqualityComparer());
         }
 
         /// <summary>
-        /// Adds or overwrites the code to the given value set by its canonical URI, and will not throw if the code exists already.
+        /// Creates a new <see cref="HashValueSetDictionary"/> with the default comparer.
         /// </summary>
-        /// <param name="valueSetUri">The value set's canonical URI.</param>
-        /// <param name="code">The code to add.</param>
-        public void Set(string valueSetUri, CqlCode code)
+        public HashValueSetDictionary() : this(CqlCodeCqlComparer.DefaultCqlComparer)
         {
-            _codesByHash[GetKey(valueSetUri, code.code, code.system)] = code;
-            _codesByHash[GetKey(valueSetUri, code.code, NullCodeSystem)] = code;
-            if (!_codesInValueSet.TryGetValue(valueSetUri, out var codes))
-            {
-                codes = new HashSet<CqlCode>(_codeHasher)
-                {
-                    code
-                };
-                _codesInValueSet.Add(valueSetUri, codes);
-            }
-            else
-                codes.Add(code);
+            // nothing
         }
 
         /// <summary>
-        /// Returns <see langword="true"/> if the given code is present in the given value set.
-        /// This method ignores the code system of the codes present in <paramref name="valueSetUri"/>.
-        /// This method approaches an O(1) operation.
+        /// Adds the code to the given value set by its canonical URI.
         /// </summary>
         /// <param name="valueSetUri">The value set's canonical URI.</param>
-        /// <param name="code">The code to check.</param>
-        /// <returns><see langword="true"/> if the given code is present in the given value set.</returns>
+        /// <param name="codes">The codes to add.</param>
+        /// <exception cref="ArgumentException">If the valueset already exists in the dictionary.</exception>
+        public void Add(string valueSetUri, IEnumerable<CqlCode> codes)
+        {
+            if (_codesInValueSet.ContainsKey(valueSetUri))
+                throw new ArgumentException($"Valueset {valueSetUri} already exists in dictionary.");
+
+            var internedCodes = codes.Select(Intern);
+            _codesInValueSet.Add(valueSetUri, new InMemoryValueSet(internedCodes));
```

Comment:
I have reused the `InMemoryValueSet` instead of re-doing comparison and lookup logic in the HashValueSetDictionary. The implementation becomes a simple hash of valusets + lookup in the valueset. This makes it possible to return ValueSetFacades when necessary, not lists of codes from a shared hashtable.